File: mathevaluator/src/test/java/com/github/vatbub/mathevaluator/MathExpressionTest.java
package com.github.vatbub.mathevaluator;

/*-
 * #%L
 * math-evaluator
 * %%
 * Copyright (C) 2016 - 2018 Frederik Kammel
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */


import org.junit.After;
import org.junit.Assert;
import org.junit.Test;

/**
 * Created by Kammel on 15.06.2018.
 */
public class MathExpressionTest {

    @After
    public void cleanUp(){
        MathLiteral.reset();
    }

    @Test
    public void simpleMultiplicationTest() {
        MathExpression expression = new MathExpression("2*10");
        Assert.assertEquals(20, expression.evaluate().getValue(), 0);
    }

    @Test
    public void negativeNumberAtTheBeginningTest() {
        MathExpression expression = new MathExpression("-2*10");
        Assert.assertEquals(-20, expression.evaluate().getValue(), 0);
    }

    @Test
    public void negativeNumberAtTheEndWithoutParenthesisTest() {
        MathExpression expression = new MathExpression("1+-2");
        Assert.assertEquals(-1, expression.evaluate().getValue(), 0);
    }

    @Test
    public void simpleSubtractionTest() {
        MathExpression expression = new MathExpression("10-2");
        Assert.assertEquals(8, expression.evaluate().getValue(), 0);
    }

    @Test
    public void simpleDivisionTest() {
        MathExpression expression = new MathExpression("4/2");
        Assert.assertEquals(2, expression.evaluate().getValue(), 0);
    }

    @Test
    public void negativeParenthesisTest() {
        MathExpression expression = new MathExpression("-(2+10)");
        Assert.assertEquals(-12, expression.evaluate().getValue(), 0);
    }

    @Test
    public void chainedOperationsTest() {
        MathExpression expression = new MathExpression("5+10*4/5*2+2");
        Assert.assertEquals(11, expression.evaluate().getValue(), 0);
    }

    @Test
    public void multiplyExpressionWithNumberTest() {
        MathExpression expression = new MathExpression("(5+10)*4");
        Assert.assertEquals(60, expression.evaluate().getValue(), 0);
    }

    @Test
    public void parenthesisTest() {
        MathExpression expression = new MathExpression("(5+10)*4/5*(2+2)");
        Assert.assertEquals(3, expression.evaluate().getValue(), 0);
    }

    @Test
    public void nestedParenthesisTest() {
        MathExpression expression = new MathExpression("5+(10*(5/(4-3)+1)/3*2)");
        Assert.assertEquals(15, expression.evaluate().getValue(), 0);
    }

    @Test
    public void expressionWithSpacesTest() {
        MathExpression expression = new MathExpression("   5   +     (  10  *  (   5 / ( 4 - 3 ) + 1   ) / 3 * 2 )    ");
        Assert.assertEquals(15, expression.evaluate().getValue(), 0);
    }

    @Test
    public void simpleSquareRootTest() {
        MathExpression expression = new MathExpression("sqrt(4)");
        Assert.assertEquals(2, expression.evaluate().getValue(), 0);
    }

    @Test
    public void nestedParenthesisWithFunctionTest() {
        MathExpression expression = new MathExpression("5+(10*(5/(4-3)+1)/3*sqrt(4))");
        Assert.assertEquals(15, expression.evaluate().getValue(), 0);
    }

    @Test
    public void squareRootWithInnerExpressionTest() {
        MathExpression expression = new MathExpression("sqrt((5+10)*4+4)");
        Assert.assertEquals(8, expression.evaluate().getValue(), 0);
    }

    @Test
    public void negatedSquareRootTest() {
        MathExpression expression = new MathExpression("-sqrt(4)");
        Assert.assertEquals(-2, expression.evaluate().getValue(), 0);
    }

    @Test
    public void squareRootOfNegativeNumberTest() {
        MathExpression expression = new MathExpression("sqrt(-4)");
        Assert.assertEquals(Double.NaN, expression.evaluate().getValue(), 0);
    }

    @Test
    public void powerTest() {
        MathExpression expression = new MathExpression("5^2");
        Assert.assertEquals(25, expression.evaluate().getValue(), 0);
    }

    @Test
    public void constantTest() {
        MathExpression expression = new MathExpression("2*pi^2");
        Assert.assertEquals(2 * Math.pow(Math.PI, 2), expression.evaluate().getValue(), 0);
    }

    @Test
    public void multiplyNegatedConstantTest() {
        MathExpression expression = new MathExpression("2*-pi");
        Assert.assertEquals(2 * -Math.PI, expression.evaluate().getValue(), 0);
    }

    @Test
    public void functionInTheMiddleOfTheExpressionTest() {
        MathExpression expression = new MathExpression("5+10*sqrt(4/5)*2+2");
        Assert.assertEquals(8 * Math.sqrt(5) + 7, expression.evaluate().getValue(), 0);
    }

    @Test
    public void implicitMultiplicationOfMathExpressionWithMathExpressionTest() {
        MathExpression expression = new MathExpression("(2^2)(3*5)");
        Assert.assertEquals(60, expression.evaluate().getValue(), 0);
    }

    @Test
    public void implicitMultiplicationOfNumberWithMathExpressionTest() {
        MathExpression expression = new MathExpression("2(3*5)");
        Assert.assertEquals(30, expression.evaluate().getValue(), 0);
    }

    @Test
    public void implicitMultiplicationOfConstantWithMathExpressionTest() {
        MathExpression expression = new MathExpression("pi(3*5)");
        Assert.assertEquals(Math.PI * 15, expression.evaluate().getValue(), 0);
    }

    @Test
    public void implicitMultiplicationOfFunctionWithMathExpressionTest() {
        MathExpression expression = new MathExpression("sin(2)(3*5)");
        Assert.assertEquals(Math.sin(2) * 15, expression.evaluate().getValue(), 0);
    }

    @Test
    public void implicitMultiplicationOfMathExpressionWithConstantTest() {
        MathExpression expression = new MathExpression("(3*5)pi");
        Assert.assertEquals(Math.PI * 15, expression.evaluate().getValue(), 0);
    }

    @Test
    public void implicitMultiplicationOfMathExpressionWithFunctionTest() {
        MathExpression expression = new MathExpression("(3*5)sin(2)");
        Assert.assertEquals(15 * Math.sin(2), expression.evaluate().getValue(), 0);
    }

    @Test
    public void implicitMultiplicationOfNumberWithConstantTest() {
        MathExpression expression = new MathExpression("2pi");
        Assert.assertEquals(2 * Math.PI, expression.evaluate().getValue(), 0);
    }

    @Test
    public void implicitMultiplicationOfNumberWithFunctionTest() {
        MathExpression expression = new MathExpression("2sin(2)");
        Assert.assertEquals(2 * Math.sin(2), expression.evaluate().getValue(), 0);
    }

    @Test
    public void implicitMultiplicationOfConstantWithConstantTest() {
        MathExpression expression = new MathExpression("pipi");
        Assert.assertEquals(Math.PI * Math.PI, expression.evaluate().getValue(), 0);
    }

    @Test
    public void implicitMultiplicationOfConstantWithFunctionTest() {
        MathExpression expression = new MathExpression("pisin(2)");
        Assert.assertEquals(Math.PI * Math.sin(2), expression.evaluate().getValue(), 0);
    }

    @Test
    public void tanTest() {
        MathExpression expression = new MathExpression("tan(5)");
        Assert.assertEquals(Math.tan(5), expression.evaluate().getValue(), 0);
    }

    @Test
    public void cotTest() {
        MathExpression expression = new MathExpression("cot(5)");
        Assert.assertEquals(1 / Math.tan(5), expression.evaluate().getValue(), 0);
    }

    @Test
    public void unbalancedParenthesisTest1() {
        assertFaultyExpression("(5+10*4/5*(2+2)", "Unbalanced parenthesis");
    }

    @Test
    public void unbalancedParenthesisTest2() {
        assertFaultyExpression("(5+10)*4/5*2+2)", "Unbalanced parenthesis");
    }

    @Test
    public void emptyStringTest() {
        assertFaultyExpression("", "Empty expression");
    }

    @Test
    public void illegalCharacterTest() {
        String inputString = "52ß?uh;:";
        String expectedIllegalString = "ß?uh;:";
        assertThrowable(() -> new MathExpression(inputString), new NumberFormatException("For input string: \"" + expectedIllegalString + "\""));
    }

    @Test
    public void operatorsOnlyTest() {
        assertThrowable(() -> new MathExpression("-*^+/"), new NumberFormatException("For input string: \"*^+/\""));
    }

    @Test
    public void functionWithoutParametersTest() {
        assertThrowable(() -> new MathExpression("2*sqrt+3"), new IllegalArgumentException("Function names must be followed by parenthesis to declare the function parameters"));
    }

    @Test
    public void runtimeConstantTest() {
        MathExpression expression1 = new MathExpression("something = 2*5");
        Assert.assertEquals(10, expression1.evaluate().getValue(), 0);
        MathExpression expression2 = new MathExpression("2*something");
        Assert.assertEquals(20, expression2.evaluate().getValue(), 0);
    }

    @Test
    public void multipleEqualSignsTest() {
        assertThrowable(() -> new MathExpression("something = 2*5 = 3"), new IllegalArgumentException("Runtime constant declarations must not contain more than one equal sign"));
    }

    @Test
    public void runtimeConstantUpdateTest() {
        new MathExpression("x = 10");
        MathExpression expression2 = new MathExpression("y = 2*x");
        Assert.assertEquals(20, expression2.evaluate().getValue(), 0);

        new MathExpression("x = 5");
        Assert.assertEquals(10, expression2.evaluate().getValue(), 0);
    }

    @Test
    public void circularRuntimeConstantDefinitionTest() {
        assertThrowable(() -> new MathExpression("something = 2*something"), new NumberFormatException("For input string: \"something\""));
    }

    @Test
    public void notInstantiableConstantTest() {
        MathLiteral.registerConstant(ConstantWithNonDefaultConstructor.class);
        try {
            assertThrowable(() -> new MathExpression("5+special+8"), new RuntimeException("java.lang.InstantiationException: com.github.vatbub.mathevaluator.MathExpressionTest$ConstantWithNonDefaultConstructor"));
        } finally {
            MathLiteral.deregisterConstant(ConstantWithNonDefaultConstructor.class);
        }
    }

    @Test
    public void notInstantiableOperatorTest() {
        MathLiteral.registerOperator(OperatorWithNonDefaultConstructor.class);
        try {
            assertThrowable(() -> new MathExpression("5bla8"), new RuntimeException("java.lang.InstantiationException: com.github.vatbub.mathevaluator.MathExpressionTest$OperatorWithNonDefaultConstructor"));
        } finally {
            MathLiteral.deregisterOperator(OperatorWithNonDefaultConstructor.class);
        }
    }

    @Test
    public void notInstantiableFunctionTest() {
        MathLiteral.registerFunction(FunctionWithNonDefaultConstructor.class);
        try {
            assertThrowable(() -> new MathExpression("5*func(8)"), new RuntimeException("java.lang.InstantiationException: com.github.vatbub.mathevaluator.MathExpressionTest$FunctionWithNonDefaultConstructor"));
        } finally {
            MathLiteral.deregisterFunction(FunctionWithNonDefaultConstructor.class);
        }
    }

    @Test
    public void circularReferenceTest1() {
        MathExpression expression = new MathExpression("r = 5");
        Assert.assertEquals(5, expression.evaluate().getValue(), 0);

        assertThrowable(() -> new MathExpression("r = r + 1"), new IllegalArgumentException("Circular reference detected"));
    }

    @Test
    public void circularReferenceTest2() {
        MathExpression expression1 = new MathExpression("r = 5");
        Assert.assertEquals(5, expression1.evaluate().getValue(), 0);
        MathExpression expression2 = new MathExpression("c = r + 1");
        Assert.assertEquals(6, expression2.evaluate().getValue(), 0);

        assertThrowable(() -> new MathExpression("r = c + 1"), new IllegalArgumentException("Circular reference detected"));
    }

    @Test
    public void doubleUnaryOperatorTest1() {
        MathExpression expression = new MathExpression("7--3");
        Assert.assertEquals(10, expression.evaluate().getValue(), 0);
    }

    @Test
    public void doubleUnaryOperatorTest2() {
        MathExpression expression = new MathExpression("7++3");
        Assert.assertEquals(10, expression.evaluate().getValue(), 0);
    }

    @Test
    public void doubleUnaryOperatorTest3() {
        MathExpression expression = new MathExpression("7---3");
        Assert.assertEquals(4, expression.evaluate().getValue(), 0);
    }

    private void assertFaultyExpression(String expression, String expectedMessage) {
        assertThrowable(() -> {
            MathExpression mathExpression = new MathExpression(expression);
            mathExpression.evaluate();
        }, new IllegalArgumentException(expectedMessage));
    }

    private <T extends Throwable> void assertThrowable(Runnable delegate, T expectedException) {
        try {
            delegate.run();
            Assert.fail("Exception expected");
        } catch (AssertionError e) {
            throw e;
        } catch (Throwable e) {
            Assert.assertEquals(expectedException.getClass().getName(), e.getClass().getName());
            Assert.assertEquals(expectedException.getMessage(), e.getMessage());
        }
    }

    public static class ConstantWithNonDefaultConstructor extends Constant {
        /**
         * The only purpose of this constructor is to hide the default constructor -->
         * Destroys the reflection mechanism used to instantiate constants
         *
         * @param value dummy argument
         */
        @SuppressWarnings("unused")
        public ConstantWithNonDefaultConstructor(Number value) {
        }

        @Override
        public Number getValue() {
            return null;
        }

        @Override
        public String getFormulaRepresentation() {
            return "special";
        }
    }

    public static class OperatorWithNonDefaultConstructor extends Operator {

        /**
         * The only purpose of this constructor is to hide the default constructor -->
         * Destroys the reflection mechanism used to instantiate operators
         *
         * @param value dummy argument
         */
        @SuppressWarnings("unused")
        public OperatorWithNonDefaultConstructor(Number value) {
        }

        @Override
        public Number evaluate(Number leftNumber, Number rightNumber) {
            return null;
        }

        @Override
        public int getPriority() {
            return 0;
        }

        @Override
        public String getFormulaRepresentation() {
            return "bla";
        }
    }

    public static class FunctionWithNonDefaultConstructor extends Function {

        /**
         * The only purpose of this constructor is to hide the default constructor -->
         * Destroys the reflection mechanism used to instantiate operators
         *
         * @param value dummy argument
         */
        @SuppressWarnings("unused")
        public FunctionWithNonDefaultConstructor(Number value) {
        }

        @Override
        public int getMinNumberOfArguments() {
            return 0;
        }

        @Override
        public int getMaxNumberOfArguments() {
            return 0;
        }

        @Override
        public Number evaluateImpl() {
            return null;
        }

        @Override
        public String getFormulaRepresentation() {
            return "func";
        }
    }
}

File: mathevaluator/src/test/kotlin/com/github/vatbub/mathevaluator/MathExpressionTest.kt
package com.github.vatbub.mathevaluator

import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.lang.Math.PI
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt
import kotlin.math.tan

/**
 * Created by Kammel on 15.06.2018.
 */
class MathExpressionTest {
    @AfterEach
    fun cleanUp() {
        MathLiteral.reset()
    }

    @Test
    fun simpleMultiplicationTest() {
        val expression = "2*10".toMathExpression()
        assertEquals(20.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun negativeNumberAtTheBeginningTest() {
        val expression = "-2*10".toMathExpression()
        assertEquals(-20.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun negativeNumberAtTheEndWithoutParenthesisTest() {
        val expression = "1+-2".toMathExpression()
        assertEquals(-1.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun simpleSubtractionTest() {
        val expression = "10-2".toMathExpression()
        assertEquals(8.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun simpleDivisionTest() {
        val expression = "4/2".toMathExpression()
        assertEquals(2.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun negativeParenthesisTest() {
        val expression = "-(2+10)".toMathExpression()
        assertEquals(-12.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun chainedOperationsTest() {
        val expression = "5+10*4/5*2+2".toMathExpression()
        assertEquals(11.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun multiplyExpressionWithNumberTest() {
        val expression = "(5+10)*4".toMathExpression()
        assertEquals(60.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun parenthesisTest() {
        val expression = "(5+10)*4/5*(2+2)".toMathExpression()
        assertEquals(3.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun nestedParenthesisTest() {
        val expression = "5+(10*(5/(4-3)+1)/3*2)".toMathExpression()
        assertEquals(15.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun expressionWithSpacesTest() {
        val expression = "   5   +     (  10  *  (   5 / ( 4 - 3 ) + 1   ) / 3 * 2 )    ".toMathExpression()
        assertEquals(15.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun simpleSquareRootTest() {
        val expression = "sqrt(4)".toMathExpression()
        assertEquals(2.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun nestedParenthesisWithFunctionTest() {
        val expression = "5+(10*(5/(4-3)+1)/3*sqrt(4))".toMathExpression()
        assertEquals(15.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun squareRootWithInnerExpressionTest() {
        val expression = "sqrt((5+10)*4+4)".toMathExpression()
        assertEquals(8.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun negatedSquareRootTest() {
        val expression = "-sqrt(4)".toMathExpression()
        assertEquals(-2.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun squareRootOfNegativeNumberTest() {
        val expression = "sqrt(-4)".toMathExpression()
        assertEquals(Double.NaN, expression.evaluate().value, 0.0)
    }

    @Test
    fun powerTest() {
        val expression = "5^2".toMathExpression()
        assertEquals(25.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun constantTest() {
        val expression = "2*pi^2".toMathExpression()
        assertEquals(2 * PI.pow(2.0), expression.evaluate().value, 0.0)
    }

    @Test
    fun multiplyNegatedConstantTest() {
        val expression = "2*-pi".toMathExpression()
        assertEquals(2 * -PI, expression.evaluate().value, 0.0)
    }

    @Test
    fun functionInTheMiddleOfTheExpressionTest() {
        val expression = "5+10*sqrt(4/5)*2+2".toMathExpression()
        assertEquals(8 * sqrt(5.0) + 7, expression.evaluate().value, 0.0)
    }

    @Test
    fun implicitMultiplicationOfMathExpressionWithMathExpressionTest() {
        val expression = "(2^2)(3*5)".toMathExpression()
        assertEquals(60.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun implicitMultiplicationOfNumberWithMathExpressionTest() {
        val expression = "2(3*5)".toMathExpression()
        assertEquals(30.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun implicitMultiplicationOfConstantWithMathExpressionTest() {
        val expression = "pi(3*5)".toMathExpression()
        assertEquals(PI * 15, expression.evaluate().value, 0.0)
    }

    @Test
    fun implicitMultiplicationOfFunctionWithMathExpressionTest() {
        val expression = "sin(2)(3*5)".toMathExpression()
        assertEquals(sin(2.0) * 15, expression.evaluate().value, 0.0)
    }

    @Test
    fun implicitMultiplicationOfMathExpressionWithConstantTest() {
        val expression = "(3*5)pi".toMathExpression()
        assertEquals(PI * 15, expression.evaluate().value, 0.0)
    }

    @Test
    fun implicitMultiplicationOfMathExpressionWithFunctionTest() {
        val expression = "(3*5)sin(2)".toMathExpression()
        assertEquals(15 * sin(2.0), expression.evaluate().value, 0.0)
    }

    @Test
    fun implicitMultiplicationOfNumberWithConstantTest() {
        val expression = "2pi".toMathExpression()
        assertEquals(2 * PI, expression.evaluate().value, 0.0)
    }

    @Test
    fun implicitMultiplicationOfNumberWithFunctionTest() {
        val expression = "2sin(2)".toMathExpression()
        assertEquals(2 * sin(2.0), expression.evaluate().value, 0.0)
    }

    @Test
    fun implicitMultiplicationOfConstantWithConstantTest() {
        val expression = "pipi".toMathExpression()
        assertEquals(PI * PI, expression.evaluate().value, 0.0)
    }

    @Test
    fun implicitMultiplicationOfConstantWithFunctionTest() {
        val expression = "pisin(2)".toMathExpression()
        assertEquals(PI * sin(2.0), expression.evaluate().value, 0.0)
    }

    @Test
    fun tanTest() {
        val expression = "tan(5)".toMathExpression()
        assertEquals(tan(5.0), expression.evaluate().value, 0.0)
    }

    @Test
    fun cotTest() {
        val expression = "cot(5)".toMathExpression()
        assertEquals(1 / tan(5.0), expression.evaluate().value, 0.0)
    }

    @Test
    fun unbalancedParenthesisTest1() {
        assertFaultyExpression("(5+10*4/5*(2+2)", "Unbalanced parenthesis")
    }

    @Test
    fun unbalancedParenthesisTest2() {
        assertFaultyExpression("(5+10)*4/5*2+2)", "Unbalanced parenthesis")
    }

    @Test
    fun emptyStringTest() {
        assertFaultyExpression("", "Empty expression")
    }

    @Test
    fun illegalCharacterTest() {
        val inputString = "52ß?uh;:"
        val expectedIllegalString = "ß?uh;:"
        assertThrowable(NumberFormatException::class.java, "For input string: \"$expectedIllegalString\"") {
            inputString.toMathExpression()
        }
    }

    @Test
    fun operatorsOnlyTest() {
        assertThrowable(NumberFormatException::class.java, "For input string: \"*^+/\"") {
            "-*^+/".toMathExpression()
        }
    }

    @Test
    fun functionWithoutParametersTest() {
        assertThrowable(
            IllegalArgumentException::class.java,
            "Function names must be followed by parenthesis to declare the function parameters"
        ) {
            "2*sqrt+3".toMathExpression()
        }
    }

    @Test
    fun runtimeConstantTest() {
        val expression1 = "something = 2*5".toMathExpression()
        assertEquals(10.0, expression1.evaluate().value, 0.0)
        val expression2 = "2*something".toMathExpression()
        assertEquals(20.0, expression2.evaluate().value, 0.0)
    }

    @Test
    fun multipleEqualSignsTest() {
        assertThrowable(
            IllegalArgumentException::class.java,
            "Runtime constant declarations must not contain more than one equal sign"
        ) { "something = 2*5 = 3".toMathExpression() }
    }

    @Test
    fun runtimeConstantUpdateTest() {
        "x = 10".toMathExpression()
        val expression2 = "y = 2*x".toMathExpression()
        assertEquals(20.0, expression2.evaluate().value, 0.0)
        "x = 5".toMathExpression()
        assertEquals(10.0, expression2.evaluate().value, 0.0)
    }

    @Test
    fun circularRuntimeConstantDefinitionTest() {
        assertThrowable(
            NumberFormatException::class.java,
            "For input string: \"something\""
        ) { "something = 2*something".toMathExpression() }
    }

    @Test
    fun notInstantiableConstantTest() {
        MathLiteral.registerConstant(ConstantWithNonDefaultConstructor::class.java)
        try {
            assertThrowable(
                NoSuchMethodException::class.java,
                "com.github.vatbub.mathevaluator.MathExpressionTest\$ConstantWithNonDefaultConstructor"
            ) { "5+special+8".toMathExpression() }
        } finally {
            MathLiteral.deregisterConstant(ConstantWithNonDefaultConstructor::class.java)
        }
    }

    @Test
    fun notInstantiableOperatorTest() {
        MathLiteral.registerOperator(OperatorWithNonDefaultConstructor::class.java)
        try {
            assertThrowable(
                NoSuchMethodException::class.java,
                "com.github.vatbub.mathevaluator.MathExpressionTest\$OperatorWithNonDefaultConstructor"
            ) { "5bla8".toMathExpression() }
        } finally {
            MathLiteral.deregisterOperator(OperatorWithNonDefaultConstructor::class.java)
        }
    }

    @Test
    fun notInstantiableFunctionTest() {
        MathLiteral.registerFunction(FunctionWithNonDefaultConstructor::class.java)
        try {
            assertThrowable(
                NoSuchMethodException::class.java,
                "com.github.vatbub.mathevaluator.MathExpressionTest\$FunctionWithNonDefaultConstructor"
            ) { "5*func(8)".toMathExpression() }
        } finally {
            MathLiteral.deregisterFunction(FunctionWithNonDefaultConstructor::class.java)
        }
    }

    @Test
    fun circularReferenceTest1() {
        val expression = "r = 5".toMathExpression()
        assertEquals(5.0, expression.evaluate().value, 0.0)
        assertThrowable(
            IllegalArgumentException::class.java,
            "Circular reference detected"
        ) { "r = r + 1".toMathExpression() }
    }

    @Test
    fun circularReferenceTest2() {
        val expression1 = "r = 5".toMathExpression()
        assertEquals(5.0, expression1.evaluate().value, 0.0)
        val expression2 = "c = r + 1".toMathExpression()
        assertEquals(6.0, expression2.evaluate().value, 0.0)
        assertThrowable(
            IllegalArgumentException::class.java,
            "Circular reference detected"
        ) { "r = c + 1".toMathExpression() }
    }

    @Test
    fun doubleUnaryOperatorTest1() {
        val expression = "7--3".toMathExpression()
        assertEquals(10.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun doubleUnaryOperatorTest2() {
        val expression = "7++3".toMathExpression()
        assertEquals(10.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun doubleUnaryOperatorTest3() {
        val expression = "7---3".toMathExpression()
        assertEquals(4.0, expression.evaluate().value, 0.0)
    }

    @Test
    fun chainedExpressionRegressionTest() {
        val expression = "8+3-5-9".toMathExpression()
        assertEquals(-3.0, expression.evaluate().value, 0.0)
    }

    private fun assertFaultyExpression(expression: String, expectedMessage: String) {
        assertThrowable(
            IllegalArgumentException::class.java,
            expectedMessage
        ) {
            val mathExpression = expression.toMathExpression()
            mathExpression.evaluate()
        }
    }

    private fun <T : Throwable?> assertThrowable(
        expectedException: Class<T>,
        expectedMessage: String,
        delegate: () -> Unit
    ) {
        Assertions.assertThrows(expectedException, {
            delegate()
        }, expectedMessage)
    }

    /**
     * The only purpose of this constructor is to hide the default constructor -->
     * Destroys the reflection mechanism used to instantiate constants
     *
     * @param value dummy argument
     */
    class ConstantWithNonDefaultConstructor(value: MathNumber) : Constant("special", value)

    /**
     * The only purpose of this constructor is to hide the default constructor -->
     * Destroys the reflection mechanism used to instantiate operators
     *
     * @param value dummy argument
     */
    class OperatorWithNonDefaultConstructor(private val value: MathNumber) : Operator("bla", 0) {
        override fun evaluate(leftNumber: MathNumber, rightNumber: MathNumber): MathNumber = value
    }

    /**
     * The only purpose of this constructor is to hide the default constructor -->
     * Destroys the reflection mechanism used to instantiate operators
     *
     * @param value dummy argument
     */
    class FunctionWithNonDefaultConstructor(private val value: MathNumber) : MathFunction("func", 0, 0) {

        override fun evaluateImpl(): MathNumber = value
    }
}
